### PR TITLE
[ntuple] Allow keys larger than 1GB in RNTuple

### PR DESCRIPTION
**Pull request by silverweed**
# This Pull request:
uses `RNTuple::fMaxKeySize` to split any payload larger than that into multiple chunks, each backed by a TKey. These keys are then read back as a single object.

Depends on #15869

## Checklist:

- [x] tested changes locally
- [x] updated the docs (if necessary)

### CODE REVIEW

File: tree/ntuple/v7/test/ntuple_storage.cxx
Comment (jblomer) [inline tree/ntuple/v7/test/ntuple_storage.cxx:911]: Validate that page list size is big enough using the `ClusterGroupDescriptor`

File: tree/ntuple/v7/test/ntuple_storage.cxx
Comment (jblomer) [inline tree/ntuple/v7/test/ntuple_storage.cxx:977]: Validate footer size using the anchor

File: tree/ntuple/v7/test/ntuple_storage.cxx
Comment (jblomer) [inline tree/ntuple/v7/test/ntuple_storage.cxx:1]: Generally, I think the compressed/uncompressed tests can be combined in a single code path, e.g. loop with the two different write options.

File: tree/ntuple/v7/src/RPageStorageFile.cxx
Comment (jblomer) [inline tree/ntuple/v7/src/RPageStorageFile.cxx:166]: That is quite sophisticated. Since the PR is anyway already large, can we for now just replace the `bytesPacked >= std::numeric_limits<std::int32_t>::max()` by `bytesPacked >= fOptions->GetMaxKeySize()` and make any further optimization subject to a follow-up PR.

**Review by jblomer — CHANGES_REQUESTED** [review]
Great work!

File: tree/ntuple/v7/inc/ROOT/RNTupleWriteOptions.hxx
Comment (silverweed) [reply tree/ntuple/v7/inc/ROOT/RNTupleWriteOptions.hxx:52]: It doesn't handle the `'` separator in 0x4000'0000 properly so it misindents the next line. If you prefer I can just remove the separator.

File: tree/ntuple/v7/inc/ROOT/RNTupleWriteOptions.hxx
Comment (silverweed) [reply tree/ntuple/v7/inc/ROOT/RNTupleWriteOptions.hxx:31]: Do we actually need a class or are we ok with a specific namespace? (e.g. `Internal::RNTupleWriteOptionManipulation::SetMaxKeySize`)
In general I tend to prefer free functions whenever the functionality is "obvious" as they tend to be easier to use and understand (no hidden state)

File: tree/ntuple/v7/inc/ROOT/RNTupleWriteOptions.hxx
Comment (jblomer) [reply tree/ntuple/v7/inc/ROOT/RNTupleWriteOptions.hxx:52]: Interesting! I was just curious. I leave it up to you.

File: tree/ntuple/v7/inc/ROOT/RNTupleWriteOptions.hxx
Comment (jblomer) [reply tree/ntuple/v7/inc/ROOT/RNTupleWriteOptions.hxx:31]: I guess the namespace would work, too. I'd still have a preference for a class (can be stateless and with static methods) because then there is for sure only one place where all the setters are defined. But in the end that's up to you.

Comment (silverweed) [reply tree/ntuple/v7/inc/ROOT/RNTupleWriteOptions.hxx]: In the end I went with a class because it's also easier to friend-declare than separate functions

File: tree/ntuple/v7/inc/ROOT/RNTupleWriteOptions.hxx
Comment (pcanal) [inline tree/ntuple/v7/inc/ROOT/RNTupleWriteOptions.hxx:57]: Does:
```suggestion
   // clang-format off
   static constexpr std::uint64_t kDefaultMaxKeySize = 0x4000'0000; // 1 GiB
   // clang-format on

```
work? (and would be clearer as it would be hiding the problematic code rather than some 'random' later code)

Comment (silverweed) [reply tree/ntuple/v7/inc/ROOT/RNTupleWriteOptions.hxx:57]: I just tried and sadly no... it indents it this way:

```c++
   // clang-format off
   static constexpr std::uint64_t kDefaultMaxKeySize = 0x4000'0000; // 1 GiB
      // clang-format on

      friend Internal::RNTupleWriteOptionsManip;

```

File: tree/ntuple/v7/src/RMiniFile.cxx
Comment (silverweed) [reply tree/ntuple/v7/src/RMiniFile.cxx:1249]: Saving it in a local potentially avoids extra loads that the compiler may insert due to the many functions with side-effects on the `this` object...although it is arguably a negligible win since we're doing lots of I/O here.

File: tree/ntuple/v7/test/ntuple_storage.cxx
Comment (silverweed) [reply tree/ntuple/v7/test/ntuple_storage.cxx:1]: Are you sure? I'm seeing that in gtest.h their own example is: (line 1840)
```
// Examples:
//
//   EXPECT_NE(Foo(), 5);
//   EXPECT_EQ(a_pointer, NULL);
//   ASSERT_LT(i, array_size);
//   ASSERT_GT(records.size(), 0) << "There is no record left.";
```

Comment (jblomer) [reply tree/ntuple/v7/test/ntuple_storage.cxx:1]: Hm, try it out then. Last time I checked, it was still "expected value" first.

File: tree/ntuple/v7/src/RMiniFile.cxx
Comment (jblomer) [inline tree/ntuple/v7/src/RMiniFile.cxx:1202]: Looking at this: if we read a version 4 RNTuple (without max key size) in the mini file, wouldn't we read the checksum as max key size?

File: tree/ntuple/v7/test/ntuple_storage.cxx
Comment (jblomer) [inline tree/ntuple/v7/test/ntuple_storage.cxx:836]: There are a number of compile errors at this place, according to github actions. Just making sure that this is a false positive.

**Review by jblomer — CHANGES_REQUESTED** [review]
Very nice! Some minor remaining comments before we should merge this.

File: tree/ntuple/v7/src/RMiniFile.cxx
Comment (silverweed) [reply tree/ntuple/v7/src/RMiniFile.cxx:1202]: True; how should be go about this? Should we declare version 4 unsupported or have some sort of manual schema evolution in RMiniFile?

File: tree/ntuple/v7/test/ntuple_storage.cxx
Comment (silverweed) [reply tree/ntuple/v7/test/ntuple_storage.cxx:1]: The output seems pretty symmetric to me:
```
Expected equality of these values:
  data[dataSize - 42]
    Which is: '\x99' (153)
  0x11
    Which is: 17
```

**Review by jblomer — APPROVED** [review]
Great! LGTM, the Windows error should be understood before merging.

File: tree/ntuple/v7/test/ntuple_compat.cxx
Comment (jblomer) [inline tree/ntuple/v7/test/ntuple_compat.cxx:209]: Could this be the Windows issue:
```suggestion
      FILE *f = fopen(fileGuard.GetPath().c_str(), "wb+");
```

**Pull request merged**
